google-cloud-pubsublite/src/test/java/com/google/cloud/pubsublite/internal/wire/PartitionCountWatchingPublisherTest.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.pubsublite.internal.wire;
+
+import static org.junit.Assert.assertThrows;
+import static org.mockito.Mockito.*;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.StatusCode;
+import com.google.cloud.pubsublite.*;
+import com.google.cloud.pubsublite.internal.ApiExceptionMatcher;
+import com.google.cloud.pubsublite.internal.CheckedApiException;
+import com.google.cloud.pubsublite.internal.Publisher;
+import com.google.cloud.pubsublite.internal.RoutingPolicy;
+import com.google.cloud.pubsublite.internal.testing.FakeApiService;
+import com.google.protobuf.ByteString;
+import java.time.Duration;
+import java.util.function.Consumer;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.*;
+
+@RunWith(JUnit4.class)
+public class PartitionCountWatchingPublisherTest {
+  abstract static class FakePublisher extends FakeApiService implements Publisher<PublishMetadata> {
+    @Override
+    public ApiFuture<PublishMetadata> publish(Message message) {
+      return ApiFutures.immediateFuture(PublishMetadata.of(Partition.of(0), Offset.of(0)));
+    }
+  }
+
+  static class FakeConfigWatcher extends FakeApiService implements PartitionCountWatcher {
+    public final Consumer<Long> consumer;
+    private final long initialCount;
+
+    protected FakeConfigWatcher(Consumer<Long> consumer, long initialCount) {
+      this.consumer = consumer;
+      this.initialCount = initialCount;
+    }
+
+    @Override
+    protected void doStart() {
+      consumer.accept(initialCount);
+      super.doStart();
+    }
+  }
```

Comment:
I think this actually adds complexity over

```
static class FakeConfigWatcher extends FakeApiService implements PartitionCountWatcher {};
@Spy FakeConfigWatcher watcher;
Consumer<Long> leakedConsumer;
```

Reply:
Changed